enhancements/machine-api/control-plane-machineset-vertical-scaling.md

```diff
@@ -0,0 +1,348 @@
+---
+title: control-plane-machineset-vertical-scaling
+authors:
+  - "@bergmannf"
+reviewers:
+  - @"JoelSpeed"
+approvers:
+  - @"JoelSpeed"
+api-approvers: 
+  - None
+creation-date: 2023-10-30
+last-updated: 2023-10-30
+tracking-link:
+  - https://issues.redhat.com/browse/OSD-15261
+see-also: []
+replaces: []
+superseded-by: []
+---
+
+# Control plane machineset automatic vertical scaling
+
+## Summary
+
+To allow control plane nodes to remain at an adequate size for its cluster, this
+proposal introduces new configuration for the control plane machineset operator
+to allow it to make automated scaling decisions for control plane node sizing.
+
+## Motivation
+
+During normal operation of an OpenShift cluster, as worker node count and
+workloads increase, it becomes necessary to increase resources available to the
+control plane nodes.
+
+During SRE operations of the managed OpenShift product, there was already a
+trigger indentified, when these increases are required. Instead of requiring the
+adjustments to be performed by hand, the operator should be able to
+automatically determine if a control plane node size change is required.
+
+On the other hand, it should also be possible to determine that control plane
+nodes have become too large for a cluster, because load has decreased. In this
+case the operator should also be able to automatically decide to scale the
+control plane down again.
+
+With the automatic increase of the control plane size, cluster performance can
+remain high, while automated decrease of the size will ensure that spend for
+users in a cloud environment is only has high as needed.
+
+### User Stories
+
+* As an OpenShift administrator, I want to know that the platform will always
+  have enough resources available to it's control plane nodes without manual
+  intervention.
+* As an OpenShift administrator, I want to know that the platform can reduce
+  spending on control plane nodes if they are oversized for the current cluster
+  load.
+* As an OpenShift administrator, I want to be able to control when sizing up and
+  down occurs, to adjust it to workloads on my cluster that might be burstable
+  to prevent the automatic scaling from occurring to often.
+
+### Goals
+
+* Allow automatic vertical scaling of control plane nodes.
```

Comment:
As an alternative, have we investigated whether horizontal scaling would work? This strikes me as potentially easier from a scaling perspective, though we would need to understand the implications for the openshift control plane which, up to this point, has only supported 3 nodes

Reply:
This is based on our workflow as SREP, where we never scale horizontally to adjust for increased load but only vertically.
So we know that vertical scaling will accommodate increased cluster load, but I am not aware if we have data about horizontal scaling as a working alternative.

Reply:
Horizontal scaling _should_ solve the issue. IIUC the issue tends to be that the API servers start getting hotter when there's an increased number of nodes/pods etc. Increasing the number of nodes horizontally would spread that load out and should therefore help.

That said, it's not currently supported by OpenShift, but, I don't really know the history of why. It may be worth asking PM/eng in a wider scope about the gaps and why we don't support horizontally scaling the control plane.

Reply:
With etcd, scaling from 3 to 4 members raises the quorum requirement from 2 to 3 (because 2 of 4 isn't  a strict majority in a split-network situation).  So you can still only accept a single member failing, and now that you have four members that could each fail, the odds of having two down at any time is higher.  So while you might handle read load better, your uptime will suffer, and folks generally consider uptime to be more important.

Scaling from 3 to 5 also raises the quorum from 2 to 3, but because you have five nodes, you can now lose two simultaneously while retaining quorum, which helps with uptime.  But because Raft requires sharing to a majority of members before committing, the amount of etcd-to-etcd network traffic grows and [write throughput decreases][1].  Most folks want high write performance.  And unless two member failures are very close together, the ControlPlaneMachineSet controller will be able to provision replacement members, and it's very rare to need the redundancy that a 5-member etcd cluster delivers.

However, that constraint is about etcd write speed, while a lot of cluster load is watches and lists and other reads.  It may be possible to scale read-only etcd members, or to leave etcd at 3 but scale Kube API servers, or similar.  But maybe there are reasons that wouldn't work.  And even if it is possible to get something like that working, it would be a fairly large architecture pivot, while we have the CPMS tooling in place today, and only lack the bit that tweaks the instance type CPMS is being asked to roll out.

[1]: https://cloud.redhat.com/blog/a-guide-to-etcd

Reply:
I think that's useful context @wking and, if there isn't an appetite within OCP to scale beyond 3 control planes (now that I know etcd operator can handle more than 3 replicas with learners), then it's a good reason to use vertical scaling. I just want to make sure this context gets into the doc so that others who see horizontal scaling dismissed understand why